### PR TITLE
Add initial Python unit/performance tests for TIFF loader module

### PR DESCRIPTION
**Pull request by gigony**
The main change is for addressing parts of #38 and #61.

- Add initial unit/perf test cases for TIFF loader
- Add license texts for python tests

This also has the following minor changes:
- Fix some typo in the doc.
- Remove cupy_rc channel in README.md
- Update name in Github templates
- Update calculation of patch_down_count

And build/doc related changes
- Update build script and contributing guide
- Fix broken links in docs
- Sort import statements by using isort

```bash
./run test                             # execute all tests
./run test python                      # execute all python tests
./run test python unit                 # execute all python unit tests
./run test python unit skimage         # execute all python unit tests in `skimage` module
./run test python unit clara           # execute all python unit tests in `clara` module
./run test python performance          # execute all python performance tests
```

Current test folder structure looks like below:

```bash
python/cucim/tests/
├── __init__.py
├── conftest.py
├── fixtures
│   └── testimage.py
├── integration
│   ├── __init__.py
│   └── clara
│       └── __init__.py
├── performance
│   ├── __init__.py
│   └── clara
│       ├── __init__.py
│       └── test_read_region_memory_usage.py
├── test_cucim.py
├── unit
│   ├── __init__.py
│   └── clara
│       ├── __init__.py
│       ├── test_image_cache.py
│       ├── test_load_image.py
│       ├── test_load_image_metadata.py
│       └── test_tiff_read_region.py
└── util
    ├── gen_image.py
    └── gen_tiff.py
```

### CODE REVIEW

File: ci/gpu/build.sh [hunk 1]
```diff
@@ -52,7 +52,9 @@ gpuci_logger "Activate conda env"
 conda activate rapids
 
 gpuci_logger "Install dependencies"
-gpuci_conda_retry install -y -c rapidsai-nightly \
+conda conda list --revisions
+
+gpuci_conda_retry install -y -c conda-forge -c rapidsai-nightly \
```
Comment (jakirkham) [inline ci/gpu/build.sh:57]:
```suggestion
gpuci_conda_retry install -y -c rapidsai-nightly \
```

File: ci/gpu/build.sh [hunk 2]
```diff
@@ -82,8 +84,14 @@ gpuci_conda_retry build -c ${LIBCUCIM_BLD_PATH} -c conda-forge -c rapidsai-night
...
-gpuci_conda_retry install -y -c ${LIBCUCIM_BLD_PATH} -c ${CUCIM_BLD_PATH} -c rapidsai-nightly \
-    "rapids-build-env=$MINOR_VERSION.*" \
+
+conda conda list --revisions
+
+gpuci_conda_retry install -y \
+    --override-channels \
+    -c ${LIBCUCIM_BLD_PATH} \
+    -c ${CUCIM_BLD_PATH} \
+    -c conda-forge \
```
Comment (jakirkham) [inline ci/gpu/build.sh:94]:
```suggestion
gpuci_conda_retry install -y -c ${LIBCUCIM_BLD_PATH} -c ${CUCIM_BLD_PATH} -c rapidsai-nightly \
     "rapids-build-env=$MINOR_VERSION.*" \
```

File: ci/gpu/build.sh [hunk 3]
```diff
@@ -109,6 +117,5 @@ else
     cat /etc/hosts
 
     gpuci_logger "Python py.test for cuCIM"
-    cd $WORKSPACE/python/cucim
-    py.test --cache-clear -v --ignore-glob . --rootdir=src
+    ./run test python all
```
Comment (jakirkham) [inline ci/gpu/build.sh:120]: Is this still needed?

**Review by jakirkham — COMMENTED** [review]
Thanks Gigon! 😄

Had a couple suggestions about CI changes that I think we can now drop as well as one question

File: ci/gpu/build.sh (hunk 3 above)
Comment (gigony) [reply ci/gpu/build.sh:120]: Yes, it is still needed. I have integrated running test cases from both skimage and clara through `run` script.

**Review by ajschmidt8 — APPROVED** [review]
Approving `ops-codeowner` file changes

**Comment by gigony** [comment]
@jakirkham now it is building. Could you please help review this?

File: python/cucim/tests/unit/clara/test_image_cache.py [hunk 4]
```diff
@@ -0,0 +1,216 @@
...
+def test_preferred_memory_capacity(testimg_tiff_stripe_32x24_16_jpeg):
+    from cucim import CuImage
+    from cucim.clara.cache import preferred_memory_capacity
+
+    img = CuImage(testimg_tiff_stripe_32x24_16_jpeg)
+
+    # same with `img.resolutions["level_dimensions"][0]`
+    image_size = img.size('XY')  # 32x24
+    tile_size = img.resolutions['level_tile_sizes'][0]  # 16x16
+    patch_size = (tile_size[0] * 2, tile_size[0] * 2)
```
Comment (grlee77) [inline python/cucim/tests/unit/clara/test_image_cache.py:93]:
```suggestion
    patch_size = (tile_size[0] * 2, tile_size[1] * 2)
```

tile_size is a 2-tuple?

File: python/cucim/tests/unit/clara/test_tiff_read_region.py [hunk 5]
```diff
@@ -0,0 +1,157 @@
...
+                    (tile_width // 2, tile_height // 2)),  # right-bottom
+                   ]
+    for (start_pos, size) in region_list:
+        cucim_arr = np.asarray(cucim_img.read_region(start_pos, size))
+
+        # Not channel values are zero, so we need to check that.
+        channel_value_count = np.apply_along_axis(np.count_nonzero, 2,
+                                                  cucim_arr)
+        count_all_zero = np.count_nonzero(channel_value_count == 0)
+        assert count_all_zero == 0
```
Comment (grlee77) [inline python/cucim/tests/unit/clara/test_tiff_read_region.py:46]:
```suggestion
        channel_value_count = np.count_nonzero(cucim_arr, axis=2)
        assert np.all(channel_value_count > 0)
```
should be equivalent (avoid `apply_along_axis` which is usually pretty inefficient)

File: python/cucim/tests/unit/clara/test_tiff_read_region.py [hunk 6]
```diff
@@ -0,0 +1,157 @@
...
+                    (tile_width, tile_height)),  # left bottom
+                   ((width - (tile_width // 2), height - (tile_height // 2)),
+                    (tile_width, tile_height)),  # right bottom
+                   ]
+
+    for (start_pos, size) in region_list:
+        cucim_arr = np.asarray(cucim_img.read_region(start_pos, size))
+        # Not channel values are zero, so we need to check that.
+        channel_value_count = np.apply_along_axis(np.count_nonzero, 2,
+                                                  cucim_arr)
```
Comment (grlee77) [inline python/cucim/tests/unit/clara/test_tiff_read_region.py:69]:
```suggestion
        channel_value_count = np.count_nonzero(cucim_arr, axis=2)
```

File: python/cucim/tests/unit/clara/test_tiff_read_region.py [hunk 7]
```diff
@@ -0,0 +1,157 @@
...
+                    (tile_width, tile_height)),  # left top (outside)
+                   ((width + tile_height, height + tile_height),
+                    (tile_width, tile_height)),  # right bottom (outside)
+                   ]
+
+    for (start_pos, size) in region_list:
+        cucim_arr = np.asarray(cucim_img.read_region(start_pos, size))
+        # All channel values should be zero, so we need to check that.
+        channel_value_count = np.apply_along_axis(np.count_nonzero, 2,
+                                                  cucim_arr)
```
Comment (grlee77) [inline python/cucim/tests/unit/clara/test_tiff_read_region.py:91]:
```suggestion
        channel_value_count = np.count_nonzero(cucim_arr, axis=2)
```

File: python/cucim/tests/unit/clara/test_tiff_read_region.py [hunk 8]
```diff
@@ -0,0 +1,157 @@
...
+
+    level_count = cucim_img.resolutions['level_count']
+    assert level_count == 6
+
+    start_pos, size = ((0, 0), (256, 256))
+    for level in range(level_count):
+        cucim_arr = np.asarray(cucim_img.read_region(start_pos, size, level))
+        # Not channel values are zero, so we need to check that.
+        channel_value_count = np.apply_along_axis(np.count_nonzero, 2,
+                                                  cucim_arr)
```
Comment (grlee77) [inline python/cucim/tests/unit/clara/test_tiff_read_region.py:117]:
```suggestion
        channel_value_count = np.count_nonzero(cucim_arr, axis=2)
```

File: python/cucim/tests/unit/clara/test_tiff_read_region.py [hunk 9]
```diff
@@ -0,0 +1,157 @@
...
+    #  'descr': [('', '|u1')], 'typestr': '|u1',
+    #  'shape': (24, 32, 3), 'version': 3}
+    assert array_interface['data'][0] is not None
+    assert not array_interface['data'][1]
+    assert array_interface['strides'] is None
+    assert array_interface['descr']
+    assert array_interface['shape'] == tuple(whole_img.shape)
+    assert array_interface['version'] == 3
+
+    def test_cuda_array_interface_support(testimg_tiff_stripe_32x24_16_jpeg):
```
Comment (grlee77) [inline python/cucim/tests/unit/clara/test_tiff_read_region.py:140]: Doesn't this whole function need to be un-indented? why is it nested within `test_array_interface_support`?

File: run [hunk 10]
```diff
@@ -656,6 +656,127 @@ copy_data_() {
...
+                run_command pip3 install --user ${packages_to_install}
+            fi
+        fi
+    fi
+    hash -r
+}
+
+install_python_test_deps_() {
+    # Package name has '<package name>[:<conda package name>]' format.
+    check_install_python_dep_ pytest pytest-cov pytest-lazy-fixture psutil GPUtil tifffile imagecodecs openslide-python
```
Comment (grlee77) [inline run:733]: may want to keep these in a separate `requirements-test.txt` file or similar? If this is the only place they are used, then it is probably not a big deal either way.

**Review by grlee77 — CHANGES_REQUESTED** [review]
Hi Gigon, this looks mostly good to me. I left just some minor comments.

The style of passing the image parameters as a string with `:` separator seems a little un-Pythonic, but this is only for test code, not part of the public API, so I don't have a problem with it.

I have not yet tried running this code, but can try that on the 5th or 6th.

File: run (hunk 10 above)
Comment (gigony) [reply run]: That's good idea! Will also use fixed versions of the packages.

**Review by gigony — COMMENTED** [review]
Thank you @grlee77 for the all valuable review! Please help review the updated code.

**Review by grlee77 — APPROVED** [review]
Thanks, the test dependency installation looks much cleaner now in the `run` script.

It would be nice to also update the developer docs to mention how the tests should be run (in CONTRIBUTING.md?)

**Comment by gigony** [comment]
@grlee77 Addressed your comments.
Please help review them.

The following commits are added:
- [Update build script and contributing guide](https://github.com/rapidsai/cucim/pull/62/commits/866b6071a19eb6590511702aa2770efc7b9cbe9c)
- [Fix broken links in docs](https://github.com/rapidsai/cucim/pull/62/commits/2b75648917b0151028b7429edd7e3997e99e5b0c)
- [Sort import statements by using isort](https://github.com/rapidsai/cucim/pull/62/commits/86535deb468f2c614945b304dc4acf4a6f0db0fb)

File: python/cucim/ci/bootstrap.py [hunk 11]
```diff
@@ -1,16 +1,11 @@
 #!/usr/bin/env python
 # -*- coding: utf-8 -*-
-from __future__ import absolute_import
-from __future__ import print_function
-from __future__ import unicode_literals
+from __future__ import absolute_import, print_function, unicode_literals
```
Comment (grlee77) [inline python/cucim/ci/bootstrap.py:3]: We shouldn't need the `__future__` imports here since we are Python 3.x only

File: python/cucim/setup.py [hunk 12]
```diff
@@ -1,17 +1,13 @@
 #!/usr/bin/env python
 # -*- encoding: utf-8 -*-
-from __future__ import absolute_import
-from __future__ import print_function
+from __future__ import absolute_import, print_function
```
Comment (grlee77) [inline python/cucim/setup.py:3]: We shouldn't need the `__future__` imports here since we are Python 3.x only

File: CONTRIBUTING.md [hunk 13]
```diff
@@ -96,56 +96,190 @@ CUDA/GPU requirements:
 You can obtain CUDA from [https://developer.nvidia.com/cuda-downloads](https://developer.nvidia.com/cuda-downloads).
 
 
-# Script to build cuCIM from source
+# Building and Testing cuCIM from Source
 
-### Build from Source
+First, please clone the cuCIM's repository
```
Comment (grlee77) [inline CONTRIBUTING.md:101]:
```suggestion
First, please clone cuCIM's repository
```

File: CONTRIBUTING.md [hunk 14]
```diff
@@ -96,56 +96,190 @@ CUDA/GPU requirements:
...
 ```bash
 CUCIM_HOME=$(pwd)/cucim
 git clone https://github.com/rapidsai/cucim.git $CUCIM_HOME
 cd $CUCIM_HOME
 ```
+## Local Development using Conda Environment (for gcc 9.x and nvcc 11.0.x)
+
+Conda can be used to setup GCC 9.x, CUDA Toolkit (including nvcc) 11.0.x, and other dependent libraries (as shown in `./conda/environments/env.yml`) for building cuCIM.
+
+Otherwise, you may need to install (such as zlib, xz, yasm) through your OS's package manager (`apt`, `yum`, and so on).
```
Comment (grlee77) [inline CONTRIBUTING.md:112]:
```suggestion
Otherwise, you may need to install dependencies (such as zlib, xz, yasm) through your OS's package manager (`apt`, `yum`, and so on).
```

**Review by grlee77 — APPROVED** [review]
I reviewed the recent commits, making only a few small suggestions

Thank you for adding the `clean` and `clean_cache` options to the run script!

File: python/cucim/ci/bootstrap.py (hunk 11 above)
Comment (gigony) [reply python/cucim/ci/bootstrap.py:3]: Thanks Greg!
oh, the file is not used in RAPIDS CI/CD but will remove it.

File: python/cucim/setup.py (hunk 12 above)
Comment (gigony) [reply python/cucim/setup.py:3]: The file is not used in RAPIDS CI/CD but will remove it.

File: conda/environments/env.yml [hunk 15]
```diff
@@ -16,8 +16,12 @@ dependencies:
   - scipy
   - python=3.8
   - cudatoolkit=11.0
+  - cudatoolkit-dev=11.0
```
Comment (jakirkham) [inline conda/environments/env.yml:19]: Why is this needed?

File: conda/environments/env.yml [hunk 16]
```diff
@@ -16,8 +16,12 @@ dependencies:
...
+  - gxx_linux-64=9
```
Comment (jakirkham) [inline conda/environments/env.yml:20]: Should we add the `cxx-compiler` package? This will handle the activation of the C++ compiler as well

Do we need the `c-compiler` too?

**Review by jakirkham — COMMENTED** [review]
Thanks Gigon! 😄

This looks pretty good. Also like the addition of the `clean` command 🙂

Had a couple minor questions about the environment file

File: conda/environments/env.yml (hunk 15 above)
Comment (gigony) [reply conda/environments/env.yml:19]: In Conda package, cudatoolkit package doesn't include `nvcc` binary, but nvcc is needed to build the project so added it. Does it make sense? Any other solution to do that? :)

File: conda/environments/env.yml (hunk 16 above)
Comment (gigony) [reply conda/environments/env.yml:20]: If I remember correctly, Installing `- gxx_linux-64=9` also installed gcc_linux-64=9.

I am not sure how cxx-compiler/c-compiler package works. would it be better to use it?

File: conda/environments/env.yml (hunk 15 above)
Comment (jakirkham) [reply conda/environments/env.yml:19]: Yeah it doesn't. We are working on a longer term solution ( https://github.com/conda-forge/cudatoolkit-feedstock/issues/62 )

However this package was added for historical reasons to help users run the CUDA installer. I don't think we should be using it

File: conda/environments/env.yml (hunk 16 above)
Comment (jakirkham) [reply conda/environments/env.yml:20]: Yeah switching to `cxx-compiler` & `c-compiler` would preferable. They operate the same way as when Conda-Build would use them

File: conda/environments/env.yml (hunk 15 above)
Comment (gigony) [reply conda/environments/env.yml:19]: @jakirkham for now, what would be a good solution until the long-term solution is available?
Currently, without installing cudatoolkit-dev, it would try to find system's nvcc if available which cannot be isolated in the environment. Do you think we can keep the change(`cudatoolkit-dev`) until the long-term solution is available?

File: conda/environments/env.yml (hunk 16 above)
Comment (gigony) [reply conda/environments/env.yml]: Updated to use cxx-compiler and c-compiler that install `gxx_linux-64=9`.
Thanks @jakirkham!

File: conda/environments/env.yml (hunk 15 above)
Comment (jakirkham) [reply conda/environments/env.yml:19]: Is it ok if we just rely on the system one for now? Or are there other issues with that?

Should add that others may just be using our prebuilt packages. So think this is more a question of what works for us

File: conda/environments/env.yml (hunk 16 above)
Comment (jakirkham) [reply conda/environments/env.yml]: Thanks Gigon :)

File: conda/environments/env.yml (hunk 15 above)
Comment (gigony) [reply conda/environments/env.yml:19]: > Is it ok if we just rely on the system one for now? Or are there other issues with that?
I thought `env.yml` is for setting up 'development environment' to build cuCIM locally from the source code and it is not for the runtime environment (it already has more than needed for just running cuCIM. e.g., yasm/compilers/cmake/flake8 etc).
 
I don't think all users (developer who want to contribute here) have the same nvcc version needed (in this case, CUDA 11.0) in the system so I think `cudatoolkit-dev` with the same version of `cudatoolkit` is needed.

May be we need to separate out `env.yml` -- one for dev, one for runtime.

**Comment by gigony** [comment]
Hi @jakirkham, can this PR be merged or do I need to address others?

**Review by jakirkham — APPROVED** [review]

**Comment by jakirkham** [comment]
@gpucibot merge

**Comment by jakirkham** [comment]
Thanks Gigon! 😄
